src/lib.rs

```diff
@@ -153,30 +153,54 @@ impl Cifs {
         -> Result<Vec<(String, NotifyAction)>, Error>
     {
         // sub-command we want to run via SMB transact
-        let cmd = subcmd::NotifySetup::new(handle.fid, subcmd::NotifyMode::all(), false);
+        let cmd = trans::NotifySetup::new(handle.fid, trans::NotifyMode::all(), false);
         // get sub-command response via transact
-        let reply: subcmd::Notification = self.transact(handle.tid, cmd).await?;
+        let reply: trans::Notification = self.transact(handle.tid, cmd).await?;
 
         Ok(reply.changes)
     }
 
 
+    pub async fn list(&mut self, share: &Share, path: &str)
+        -> Result<Vec<DirInfo>, Error>
+    {
+        let search = FileAttr::HIDDEN | FileAttr::SYSTEM | FileAttr::DIRECTORY;
+        let cmd = trans2::subcmd::FindFirst2::new(sanitize_path(path), search);
+        let reply: trans2::subreply::FindFirst2 = self.transact2(share.tid, cmd).await?;
+
+        // are we done yet?
+        if reply.end {
+            return Ok(reply.info);
+        }
+
+        let sid = reply.sid;
+        let mut last = reply.info.last().unwrap().filename.clone();
+        let mut result = reply.info;
+
+        loop {
+            let cmd = trans2::subcmd::FindNext2::new(sid, last);
+            let mut reply: trans2::subreply::FindNext2 = self.transact2(share.tid, cmd).await?;
+
+            last = reply.info.last().unwrap().filename.clone();
+            result.append(&mut reply.info);
+
+            if reply.end {
+                break;
+            }
```

Comment:
Since this sequence of calling `transact2` and checking for `reply.end` happens twice, it might make sense to add some information about the semantics here. If I were to guess the first `reply.end` check is only there for the special case that there are no files in the directory at all? Nonetheless it seems weird that in one case `end` has to be checked before accessing the filename and in the other one it's after retrieving the filename...

Reply:
i cleaned it up a bit..